File: apps/relay-go/pkg/txbuilder/sighash.go
package txbuilder

import (
	"fmt"
)

var zero32Sig = make([]byte, 32)

// SighashInputs carries the per-input data needed to compute the
// BIP-143 sighash that BSV consensus verifies.
type SighashInputs struct {
	Tx         BsvTransaction
	InputIdx   int
	PrevScript []byte
	PrevValue  uint64
	HashType   uint32
}

func isBaseSighash(t uint32, base uint32) bool { return (t & 0x1f) == base }

func hashPrevouts(tx BsvTransaction, hashType uint32) []byte {
	if hashType&SighashAnyoneCanPay != 0 {
		return zero32Sig
	}
	parts := [][]byte{}
	for _, in := range tx.Inputs {
		parts = append(parts, in.PrevTxid, EncU32LE(in.PrevVout))
	}
	return sha256d(concat(parts...))
}

func hashSequence(tx BsvTransaction, hashType uint32) []byte {
	if hashType&SighashAnyoneCanPay != 0 {
		return zero32Sig
	}
	if isBaseSighash(hashType, SighashNone) || isBaseSighash(hashType, SighashSingle) {
		return zero32Sig
	}
	parts := [][]byte{}
	for _, in := range tx.Inputs {
		parts = append(parts, EncU32LE(in.Sequence))
	}
	return sha256d(concat(parts...))
}

func hashOutputs(tx BsvTransaction, inputIdx int, hashType uint32) []byte {
	if isBaseSighash(hashType, SighashAll) {
		parts := [][]byte{}
		for _, o := range tx.Outputs {
			parts = append(parts,
				EncU64LE(o.Value),
				EncVarint(uint64(len(o.LockingScript))),
				o.LockingScript,
			)
		}
		return sha256d(concat(parts...))
	}
	if isBaseSighash(hashType, SighashSingle) && inputIdx < len(tx.Outputs) {
		o := tx.Outputs[inputIdx]
		return sha256d(concat(
			EncU64LE(o.Value),
			EncVarint(uint64(len(o.LockingScript))),
			o.LockingScript,
		))
	}
	return zero32Sig
}

// ComputeSighash returns the BIP-143 sighash for the named input.
// SIGHASH_FORKID is mandatory on BSV; the function rejects any
// hashType missing it.
func ComputeSighash(in SighashInputs) ([]byte, error) {
	if in.HashType&SighashForkID == 0 {
		return nil, fmt.Errorf("ComputeSighash: BSV requires SIGHASH_FORKID")
	}
	if in.InputIdx < 0 || in.InputIdx >= len(in.Tx.Inputs) {
		return nil, fmt.Errorf("ComputeSighash: input index %d out of range", in.InputIdx)
	}
	i := in.Tx.Inputs[in.InputIdx]
	preimage := concat(
		EncI32LE(in.Tx.Version),
		hashPrevouts(in.Tx, in.HashType),
		hashSequence(in.Tx, in.HashType),
		i.PrevTxid,
		EncU32LE(i.PrevVout),
		EncVarint(uint64(len(in.PrevScript))),
		in.PrevScript,
		EncU64LE(in.PrevValue),
		EncU32LE(i.Sequence),
		hashOutputs(in.Tx, in.InputIdx, in.HashType),
		EncU32LE(in.Tx.LockTime),
		EncU32LE(in.HashType),
	)
	return sha256d(preimage), nil
}

File: packages/tx-builder/src/sighash.ts
/**
 * BSV post-Genesis BIP-143 sighash.
 *
 * Single canonical preimage format used by every modern BSV tx
 * (SIGHASH_FORKID must be set). Reference: bitcoinsv specification
 * 2017-bitcoin-cash-uahf, restored to BSV post-Genesis.
 *
 *   preimage = hashPrevouts (32)
 *           || hashSequence (32)
 *           || outpoint     (36)
 *           || scriptCode   (varint || bytes)
 *           || value        (8)
 *           || sequence     (4)
 *           || hashOutputs  (32)
 *           || lockTime     (4)
 *           || sighashType  (4)
 *
 * sighash = double-SHA-256(preimage).
 *
 * SIGHASH_NONE and SIGHASH_SINGLE variants zero out parts of the
 * preimage per the same spec. SIGHASH_ANYONECANPAY clears
 * hashPrevouts and hashSequence.
 */

import type { BsvTransaction } from './types.js';
import { SIGHASH_ALL, SIGHASH_ANYONECANPAY, SIGHASH_FORKID, SIGHASH_NONE, SIGHASH_SINGLE } from './types.js';
import { concat, encU32LE, encU64LE, encVarint } from './encode.js';

async function sha256(input: Uint8Array): Promise<Uint8Array> {
  const buf = new ArrayBuffer(input.byteLength);
  new Uint8Array(buf).set(input);
  return new Uint8Array(await crypto.subtle.digest('SHA-256', buf));
}

async function doubleSha256(input: Uint8Array): Promise<Uint8Array> {
  return sha256(await sha256(input));
}

const ZERO_32 = new Uint8Array(32);

function isBaseSighash(t: number, base: number): boolean {
  return (t & 0x1f) === base;
}

async function hashPrevouts(tx: BsvTransaction, hashType: number): Promise<Uint8Array> {
  if ((hashType & SIGHASH_ANYONECANPAY) !== 0) return ZERO_32;
  const parts: Uint8Array[] = [];
  for (const i of tx.inputs) {
    parts.push(i.prevTxid);
    parts.push(encU32LE(i.prevVout));
  }
  return doubleSha256(concat(...parts));
}

async function hashSequence(tx: BsvTransaction, hashType: number): Promise<Uint8Array> {
  if ((hashType & SIGHASH_ANYONECANPAY) !== 0) return ZERO_32;
  if (isBaseSighash(hashType, SIGHASH_NONE)) return ZERO_32;
  if (isBaseSighash(hashType, SIGHASH_SINGLE)) return ZERO_32;
  const parts: Uint8Array[] = [];
  for (const i of tx.inputs) parts.push(encU32LE(i.sequence));
  return doubleSha256(concat(...parts));
}

async function hashOutputs(tx: BsvTransaction, inputIdx: number, hashType: number): Promise<Uint8Array> {
  if (isBaseSighash(hashType, SIGHASH_ALL)) {
    const parts: Uint8Array[] = [];
    for (const o of tx.outputs) {
      parts.push(encU64LE(o.value));
      parts.push(encVarint(o.lockingScript.byteLength));
      parts.push(o.lockingScript);
    }
    return doubleSha256(concat(...parts));
  }
  if (isBaseSighash(hashType, SIGHASH_SINGLE) && inputIdx < tx.outputs.length) {
    const o = tx.outputs[inputIdx]!;
    return doubleSha256(concat(encU64LE(o.value), encVarint(o.lockingScript.byteLength), o.lockingScript));
  }
  return ZERO_32;
}

/**
 * Compute the BIP-143 sighash for `tx.inputs[inputIdx]` spending the
 * UTXO whose locking script is `prevScript` and whose value is
 * `prevValue` satoshis. The returned 32-byte digest is what gets
 * ECDSA-signed; the signature is then concatenated with the 1-byte
 * `hashType` and pushed onto the unlocking script.
 *
 * SIGHASH_FORKID MUST be set on `hashType` (BSV-mandatory).
 */
export async function computeSighash(args: {
  readonly tx: BsvTransaction;
  readonly inputIdx: number;
  readonly prevScript: Uint8Array;
  readonly prevValue: bigint;
  readonly hashType: number;
}): Promise<Uint8Array> {
  const { tx, inputIdx, prevScript, prevValue, hashType } = args;
  if ((hashType & SIGHASH_FORKID) === 0) {
    throw new Error('computeSighash: BSV requires SIGHASH_FORKID');
  }
  if (inputIdx < 0 || inputIdx >= tx.inputs.length) {
    throw new Error(`computeSighash: inputIdx ${inputIdx} out of range`);
  }
  const i = tx.inputs[inputIdx]!;
  const preimage = concat(
    new Uint8Array([tx.version & 0xff, (tx.version >> 8) & 0xff, (tx.version >> 16) & 0xff, (tx.version >> 24) & 0xff]),
    await hashPrevouts(tx, hashType),
    await hashSequence(tx, hashType),
    i.prevTxid,
    encU32LE(i.prevVout),
    encVarint(prevScript.byteLength),
    prevScript,
    encU64LE(prevValue),
    encU32LE(i.sequence),
    await hashOutputs(tx, inputIdx, hashType),
    encU32LE(tx.lockTime),
    encU32LE(hashType),
  );
  return doubleSha256(preimage);
}
